text/0000-custom-dst.md

```diff
@@ -0,0 +1,476 @@
+    t as *const U
+}
+```
+
+so we do not introduce any new functions to access the pointer part
+of the thick pointer.
+
+The `DynamicallySized` trait may be implemented for a struct or union
+-- The author is of the opinion that implementing it for `enum`s is
+unlikely to be useful. That may be a future extension,
+if people are interested
+(once `<T: ?Sized>` is allowed on `enum` declarations).
+
+`DynamicallySized` will be placed into the prelude.
+
+# Drawbacks
+[drawbacks]: #drawback
+
+- More complication in the language.
+- Lack of a `Sized` type dual to these unsized types --
+  the lack of a `[u8; N]` to these types' `[u8]` is unfortunate.
+- Inability to define a custom DST safely
+- The `size_of_val` and `align_of_val` declarations are now incorrect;
+  they should take `T: DynamicallySized`, as opposed to `T: ?Sized`.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+This has been a necessary change for quite a few years.
+The only real alternatives are those which are simply different ways of writing
+this feature. We need custom DSTs.
+
+We also likely especially want to deprecate the current `?Sized` behavior of
+`size_of_val` and `align_of_val`, since people are planning on
+aborting/panicking at runtime when called on extern types. That's not great.
+([link](https://github.com/rust-lang/rfcs/pull/2310#issuecomment-384770802))
+
+# Prior art
+[prior-art]: #prior-art
+
+- [FAMs in C](https://en.wikipedia.org/wiki/Flexible_array_member)
+- [FAMs in C++](https://htmlpreview.github.io/?https://github.com/ThePhD/future_cxx/blob/master/papers/d1039.html) (unfinished proposal)
+- Existing Rust which could use this feature:
+  - [CStr](https://doc.rust-lang.org/stable/std/ffi/struct.CStr.html)
+  - [Pascal String](https://github.com/ubsan/epsilon/blob/master/src/string.rs#L11)
+  - [Bit Vector](https://github.com/skiwi2/bit-vector/blob/master/src/bit_slice.rs)
+- Other RFCs
+  - [mzabaluev's Version](https://github.com/rust-lang/rfcs/pull/709)
+  - [My Old Version](https://github.com/rust-lang/rfcs/pull/1524)
+  - [japaric's Pre-RFC](https://github.com/japaric/rfcs/blob/unsized2/text/0000-unsized-types.md)
+  - [mikeyhew's Pre-RFC](https://internals.rust-lang.org/t/pre-erfc-lets-fix-dsts/6663)
+  - [MicahChalmer's RFC](https://github.com/rust-lang/rfcs/pull/9)
+  - [nrc's Virtual Structs](https://github.com/rust-lang/rfcs/pull/5)
+  - [Pointer Metadata and VTable](https://github.com/rust-lang/rfcs/pull/2580)
+  - [Syntax of ?Sized](https://github.com/rust-lang/rfcs/pull/490)
+- [Niko's Blog on DSTs](http://smallcultfollowing.com/babysteps/blog/2014/01/05/dst-take-5/)
+
+(you will note the incredible number of RFCs on this topic -- we really need to fix this missing feature)
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- Bikeshedding names.
+- Should `Metadata` really require all of those traits?
```

Comment:
I believe this was at least partially addressed by the comment in the first code snippet in the Guide-level Explanation section:
```
    /*
        note: these are all required due to trait implementations for
          - Unpin - all pointer types
          - Copy + Send + Sync - &T, &mut T
          - Eq + Ord - *const T, *mut T
    */
    type Metadata: 'static + Copy + Send + Sync + Eq + Ord + Unpin;
```
Though this probably deserves to be explained in complete sentences in prose. Just from this, I'm not clear on why `Unpin` is required, or whether `'static` is really necessary.

Reply:
Yeah I wasn't sure whether to leave this comment on this part or the part you've cited; I don't think the guide really motivates it.

Reply:
@Ixrec because we want `Box<T> : Unpin`. `'static` is necessary for `Box<T>: 'static`; otherwise, for everything that takes `T: ?Sized`, `Box<T>` might not be `'static`, which is kinda crazy.

Reply:
`Eq + Ord` seem unfortunate, because it means e.g. vtable pointers, despite being effectively `&'static`, compare by-address instead of by-value.

But vtables are already complicated magic so this seems fine otherwise.

(okay it's still weird that raw pointers implement `Ord` even for `T::Metadata != ()`)